Add service discovery implementation for static files

doeg opened the pull request:
Signed-off-by: Sara Bee <855595+doeg@users.noreply.github.com>

ajm188 commented on go/vt/vtadmin/cluster/discovery/discovery_static_file.go:
```diff
@@ -0,0 +1,142 @@
+			gates = append(gates, g)
+		}
+	} else {
+		for _, t := range tags {
```
I think this is a bug because I didn't document what tags actually mean semantically (sorry!). Tags should be logically AND-ed, rather than OR-ed; the way this is written, if a gate has any of the tags passed in, it will be included in the result, rather than only gates that have all of them.

doeg replied:
:shake_head:

ajm188 reviewed (commented):
very nice 😊  just a couple suggestions and one mea culpa on my part and i think this is ready to merge upstream, once ..... _sigh_ the other one gets merged

ajm188 commented on go/vt/vtadmin/cluster/discovery/discovery.go:
```diff
@@ -75,4 +75,5 @@ func New(impl string, cluster string, args []string) (Discovery, error) {
 
 func init() { // nolint:gochecknoinits
 	Register("consul", NewConsul)
+	Register("static", NewStaticFile)
```
Oh, I wonder ... what do you think about "static" vs "staticfile"?

doeg replied:
i like

ajm188 replied:
err 😅 I meant the other way haha

because it's both static and sourced from a file. but i don't actually feel that strongly

doeg replied:
Ah, sorry for misunderstanding. Your request was so polite. :')

doeg commented on go/vt/vtadmin/cluster/discovery/discovery_static_file.go:
```diff
@@ -0,0 +1,126 @@
+		return nil, err
+	}
+
+	defer jsonFile.Close()
```
Oh nice, does that mean I can replace:

```go

	jsonFile, err := ioutil.ReadFile(*filePath)
	if err != nil {
		return nil, err
	}

	defer jsonFile.Close()

	b, err := ioutil.ReadAll(jsonFile)
	if err != nil {
		return nil, err
	}
```

...with:

```go
	b, err := ioutil.ReadFile(*filePath)
	if err != nil {
		return nil, err
	}
```

It seems to work fine, so! Thanks!

doeg commented on go/vt/vtadmin/cluster/discovery/discovery_static_file.go:
```diff
@@ -0,0 +1,126 @@
+		return nil, err
+	}
+
+	jsonFile, err := os.Open(*filePath)
```
Do you just mean:

```go
	filePath := flags.String("path", "", "path to the static JSON file")
	if filePath == "" {
	  return nil, errors.New("Missing path")
	}	

        if err := flags.Parse(args); err != nil {
		return nil, err
	}
```

...orrrrr something else? I thought `filePath` wouldn't be set until `flags.Parse` is called. :thinking:

ajm188 replied:
Yeah, you have to check after the call to `Parse`, but other Go error caveats (should not begin with a capital letter (weird!), should provide sufficient context, etc)

doeg commented:
@ajm188 I addressed all your feedback, but p l e a s e let me know if you think of anything else! 😎 I appreciate nit-picky feedback so don't hold back.

doeg commented on go/vt/vtadmin/cluster/discovery/discovery_static.go:
```diff
@@ -0,0 +1,161 @@
+
+	set := d.gates.byName
+
+	for _, tag := range tags {
```
I stole this from fakediscovery 😎 many thank
